### PR TITLE
style: Move border-image-repeat outside of mako.

### LINKED ISSUE #19142: Move border-image-repeat outside of mako.
As part of #19015

### PR DESCRIPTION
**Pull request by chansuke**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #19142.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19210)

### CODE REVIEW

File: components/style/values/computed/border.rs [hunk 1]
```diff
@@ -81,3 +87,59 @@ impl ToAnimatedZero for BorderCornerRadius {
...
+        BorderImageRepeat(RepeatKeyword::Repeat, RepeatKeyword::Repeat)
+    }
+}
+
+impl ToCss for BorderImageRepeat {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result
+    where
+        W: fmt::Write,
+    {
+        match (self.0, self.1) {
```
Comment (emilio) [inline components/style/values/computed/border.rs]: There's no need for `match` here, just:

```rust
let BorderImageRepeat(horizontal, vertical) = *self;
```

File: components/style/values/computed/border.rs [hunk 2]
```diff
@@ -81,3 +87,59 @@ impl ToAnimatedZero for BorderCornerRadius {
...
+            }
+            SpecifiedBorderImageRepeat::Keywords(horizontal, vertical) => {
+                BorderImageRepeat(horizontal, vertical.unwrap_or(horizontal))
+            }
+        }
+    }
+
+    #[inline]
+    fn from_computed_value(computed: &Self::ComputedValue) -> Self {
+        match (computed) {
```
Comment (emilio) [inline components/style/values/computed/border.rs]: Thid doesn't build does it?

**Review by emilio — COMMENTED** [review]
This has a couple of build errors, but it's a good start!

Let me know if I can help fixing them somehow.

**Comment by chansuke** [comment]
@emilio Thanks for the comment!!I'll fix it:)

File: components/layout/display_list_builder.rs [hunk 3]
```diff
@@ -1680,8 +1681,8 @@ impl FragmentDisplayListBuilding for Fragment {
                                                           corners.3.resolve(webrender_image.width)),
                                 // TODO(gw): Support border-image-outset
                                 outset: SideOffsets2D::zero(),
-                                repeat_horizontal: convert_repeat_mode(border_style_struct.border_image_repeat.0),
-                                repeat_vertical: convert_repeat_mode(border_style_struct.border_image_repeat.1),
+                                repeat_horizontal: convert_repeat_mode(border_style_struct.BorderImageRepeatKeyword.0),
```
Comment (emilio) [inline components/layout/display_list_builder.rs]: This is also not compiling, I think

File: components/style/values/computed/border.rs [hunk 4]
```diff
@@ -81,3 +87,50 @@ impl ToAnimatedZero for BorderCornerRadius {
...
+        BorderImageRepeat(RepeatKeyword::Repeat, RepeatKeyword::Repeat)
+    }
+}
+
+impl ToCss for BorderImageRepeat {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result
+    where
+        W: fmt::Write,
+    {
+        let BorderImageRepeat(horizontal, vertical) = *self;
```
Comment (emilio) [inline components/style/values/computed/border.rs]: Now you need to actually serialize them, keeping the code that existed.

File: components/style/values/computed/border.rs (hunk 2 above)
Comment (emilio) [reply components/style/values/computed/border.rs]: This is probably still not building, let me know if you need any help setting up the build environment or if I can help somehow, I'd be glad to :)

File: components/style/values/computed/border.rs (hunk 4 above)
Comment (chansuke) [reply components/style/values/computed/border.rs]: Ok.Thank you for the review.I'll fix it

**Comment by chansuke** [comment]
@emilio Sorry for interval.I fixed little bit but stuck...I can't solve the build error below.
- unresolved import `values::specified::BorderImageRepeat`
- failed to resolve. Could not find `BorderImageRepeat` in `specified`
- cannot find function `get_initial_specified_value` in module `border_image_repeat`

**Comment by emilio** [comment]
Sorry for the huge lag here @chansuke!

For the unresolved import, you need to export that value from `values/specified/mod.rs`.

For `get_initial_specified_value` error, you need to remove the `vector=True` bit.

Other than that it looks great!

**Comment by chansuke** [comment]
@emilio Thank you for your help!!I fixed the error.The remaining error is `missing documentation for a method` error but how can I solve this?

```
error: missing documentation for a method
  --> components/style/values/computed/border.rs:95:5
   |
95 | /     pub fn repeat() -> Self {
96 | |         BorderImageRepeat(RepeatKeyword::Repeat, RepeatKeyword::Repeat)
97 | |     }
   | |_____^
   |
note: lint level defined here
  --> components/style/values/mod.rs:9:9
   |
9  | #![deny(missing_docs)]
   |         ^^^^^^^^^^^^

error: missing documentation for an enum
   --> components/style/values/specified/border.rs:184:1
    |
184 | / pub enum BorderImageRepeat {
185 | |     /// `[ stretch | repeat | round | space ]{1,2}`
186 | |     Keywords(RepeatKeyword, Option<RepeatKeyword>),
187 | | }
    | |_^

error: aborting due to 2 previous errors
```

**Comment by CYBAI** [comment]
@chansuke You only need to write `documentation comments` for them; then the error will disappear. Documentation comments in Rust always start with `///` or `//!`.

For example, here are documentation comments for `BorderSideWidth`
https://github.com/servo/servo/blob/f7440bf1a68cd0beaf12621c9c460c6f6ddc57e3/components/style/values/specified/border.rs#L21-L32

Ref: [Rust By Example - Documentation](https://rustbyexample.com/meta/doc.html)

**Comment by chansuke** [comment]
@CYBAI Thanks!

**Comment by jdm** [comment]
```
error[E0432]: unresolved import `style::properties::longhands::computed_value`

  --> components/layout/display_list_builder.rs:61:35

   |

61 | use style::properties::longhands::computed_value::RepeatKeyword;

   |                                   ^^^^^^^^^^^^^^ Could not find `computed_value` in `longhands`

error[E0433]: failed to resolve. Use of undeclared type or module `BackgroundOrigin`

    --> components/layout/display_list_builder.rs:1449:13

     |

1449 |             BackgroundOrigin::BorderBox => {}

     |             ^^^^^^^^^^^^^^^^ Use of undeclared type or module `BackgroundOrigin`

error[E0433]: failed to resolve. Use of undeclared type or module `BackgroundOrigin`

    --> components/layout/display_list_builder.rs:1450:13

     |

1450 |             BackgroundOrigin::PaddingBox => {

     |             ^^^^^^^^^^^^^^^^ Use of undeclared type or module `BackgroundOrigin`

error[E0433]: failed to resolve. Use of undeclared type or module `BackgroundOrigin`

    --> components/layout/display_list_builder.rs:1457:13

     |

1457 |             BackgroundOrigin::ContentBox => {

     |             ^^^^^^^^^^^^^^^^ Use of undeclared type or module `BackgroundOrigin`

warning: unused import: `style::values::specified::border::RepeatKeyword as BorderImageRepeatKeyword`

  --> components/layout/display_list_builder.rs:76:5

   |

76 | use style::values::specified::border::RepeatKeyword as BorderImageRepeatKeyword;

   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

   |

   = note: #[warn(unused_imports)] on by default

error[E0609]: no field `BorderImageRepeatKeyword` on type `&style::properties::style_structs::Border`

    --> components/layout/display_list_builder.rs:1707:88

     |

1707 |                             repeat_horizontal: convert_repeat_mode(border_style_struct.BorderImageRepeatKeyword.0),

     |                                                                                        ^^^^^^^^^^^^^^^^^^^^^^^^

error[E0609]: no field `BorderImageRepeatKeyword` on type `&style::properties::style_structs::Border`

    --> components/layout/display_list_builder.rs:1708:86

     |

1708 |                             repeat_vertical: convert_repeat_mode(border_style_struct.BorderImageRepeatKeyword.1),

     |                                                                                      ^^^^^^^^^^^^^^^^^^^^^^^^

error[E0609]: no field `BorderImageRepeatKeyword` on type `&style::properties::style_structs::Border`

    --> components/layout/display_list_builder.rs:1740:92

     |

1740 |                                 repeat_horizontal: convert_repeat_mode(border_style_struct.BorderImageRepeatKeyword.0),

     |                                                                                            ^^^^^^^^^^^^^^^^^^^^^^^^

error[E0609]: no field `BorderImageRepeatKeyword` on type `&style::properties::style_structs::Border`

    --> components/layout/display_list_builder.rs:1741:90

     |

1741 |                                 repeat_vertical: convert_repeat_mode(border_style_struct.BorderImageRepeatKeyword.1),

     |                                                                                          ^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 8 previous errors

error: Could not compile `layout`.
```

**Comment by chansuke** [comment]
Sorry for the lag.I will rebase and fix the error.

**Comment by emilio** [comment]
Looks fine, but there are still Geckolib build errors (you can check them with `./mach build-geckolib` or `./mach cargo-geckolib check`)

**Comment by chansuke** [comment]
@emilio Hi.I'm blocking with these build errors below.I was wondering if you could give me some suggestions.Thanks.
```
error[E0432]: unresolved import `style::properties::longhands::border_image_repeat::computed_value::RepeatKeyword`
  --> components/layout/display_list/builder.rs:67:5

error[E0611]: field `0` of tuple-struct `style::values::computed::BorderImageRepeat` is private
    --> components/layout/display_list/builder.rs:1452:33

error[E0611]: field `1` of tuple-struct `style::values::computed::BorderImageRepeat` is private
    --> components/layout/display_list/builder.rs:1455:33

error[E0611]: field `0` of tuple-struct `style::values::computed::BorderImageRepeat` is private
    --> components/layout/display_list/builder.rs:1492:37

error[E0611]: field `1` of tuple-struct `style::values::computed::BorderImageRepeat` is private
    --> components/layout/display_list/builder.rs:1495:37
```

**Comment by CYBAI** [comment]
Hi @chansuke 
cc @emilio 

With your patch, `RepeatKeyword` will be initiated from `style::properties::longhands::border_image_repeat::computed_value`.
Thus, you will need to import it from the `enum` your created in `specified` module.

https://github.com/servo/servo/blob/005dc85a888b1c73a849bb7b4e2f7d9d22321909/components/layout/display_list/builder.rs#L67

About other error messages which is `tuple-struct style::values::computed::BorderImageRepeat is private`, it means your `struct` fields should be `pub`lic.

**Comment by chansuke** [comment]
@emilio I fixed build errors.I wonder if you could check this.

File: components/style/values/specified/border.rs [hunk 5]
```diff
@@ -171,3 +172,47 @@ impl Parse for BorderSpacing {
...
+    "space" => Space
+}
+
+/// The specified value for the `border-image-repeat` property.
+///
+/// https://drafts.csswg.org/css-backgrounds/#the-border-image-repeat
+#[derive(Clone, Debug, MallocSizeOf, PartialEq, ToCss)]
+pub enum BorderImageRepeat {
+    /// `[ stretch | repeat | round | space ]{1,2}`
+    Keywords(RepeatKeyword, Option<RepeatKeyword>),
```
Comment (emilio) [inline components/style/values/specified/border.rs]: This is just an enum with a single variant, so this can be a `struct BorderImageRepeat(RepeatKeyword, Option<RepeatKeyword>)`, right?

File: components/style/values/computed/inherited_box.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+//! Computed values for inherited box
+
+use std::fmt;
+use style_traits::ToCss;
+use values::specified::Angle;
+
+/// An angle rounded and normalized per https://drafts.csswg.org/css-images/#propdef-image-orientation
+#[allow(missing_docs)]
+#[derive(Clone, Copy, Debug, Eq, MallocSizeOf, PartialEq)]
+pub enum Orientation {
```
Comment (emilio) [inline components/style/values/computed/inherited_box.rs]: These changes are from another unrelated commit, so can just go away.

File: components/style/values/computed/border.rs [hunk 7]
```diff
@@ -81,3 +84,56 @@ impl ToAnimatedZero for BorderCornerRadius {
...
+                BorderImageRepeat(horizontal, vertical.unwrap_or(horizontal))
+            }
+        }
+    }
+
+    #[inline]
+    fn from_computed_value(computed: &Self::ComputedValue) -> Self {
+        match (computed.0, computed.1) {
+            (horizontal, vertical) => {
+                SpecifiedBorderImageRepeat::Keywords(horizontal, Some(vertical))
```
Comment (emilio) [inline components/style/values/computed/border.rs]: This can just be:

```rust
SpecifiedBorderImageRepeat::Keywords(computed.0, Some(computed.1))
```

Instead of this `match` expression.

File: components/style/values/computed/border.rs (hunk 1 above)
Comment (emilio) [reply components/style/values/computed/border.rs]: This comment still applies.

**Review by emilio — COMMENTED** [review]
Looks good! Some stuff still could be cleaner, but let me know if you just want to land this and we can tidy this up in followups :)

File: components/style/values/computed/border.rs [hunk 8]
```diff
@@ -81,3 +84,56 @@ impl ToAnimatedZero for BorderCornerRadius {
...
+impl ToComputedValue for SpecifiedBorderImageRepeat {
+    type ComputedValue = BorderImageRepeat;
+
+    #[inline]
+    fn to_computed_value(&self, _: &Context) -> Self::ComputedValue {
+        match *self {
+            SpecifiedBorderImageRepeat::Keywords(horizontal, vertical) => {
+                BorderImageRepeat(horizontal, vertical.unwrap_or(horizontal))
+            }
+        }
```
Comment (CYBAI) [inline components/style/values/computed/border.rs]: I'm wondering, when we only have one pattern, will it be better to use `if let` here?

File: components/style/values/specified/border.rs [hunk 9]
```diff
@@ -171,3 +172,47 @@ impl Parse for BorderSpacing {
...
+            Ok(h) => h,
+            Err(()) => {
+                return Err(input.new_custom_error(
+                    SelectorParseErrorKind::UnexpectedIdent(ident.clone())
+                ));
+            }
+        };
+
+        let vertical = input.try(RepeatKeyword::parse).ok();
+        Ok(BorderImageRepeat::Keywords(horizontal, vertical))
```
Comment (CYBAI) [inline components/style/values/specified/border.rs]: IMO, the original `parse` method looks easier to understand for me.
Just curious if you met any difficulty when you tried to use the original method?

```rust
let first = RepeatKeyword::parse(input)?;
let second = input.try(RepeatKeyword::parse).ok();

Ok(SpecifiedValue(first, second))
```

**Comment by emilio** [comment]
Sorry I didn't notice the followup commit @chansuke! I rebased, fixed the nits, and fixed the initial value, which you had changed to `repeat` instead of `stretch`, and opened #19924 with it.

Thanks a lot for the patch!

**Comment by chansuke** [comment]
@emilio @CYBAI Thank you so much for taking the time!
